File: map.js
import puppeteer from "puppeteer-extra";
import StealthPlugin from "puppeteer-extra-plugin-stealth"
import ExcelJS from "exceljs"
import path from "path"
import * as XLSX from "xlsx"
import AdblockerPlugin from "puppeteer-extra-plugin-adblocker"

puppeteer.use(StealthPlugin());
puppeteer.use(AdblockerPlugin({ blockTrackers: true }));

(async ()=> {
    try {
        const browser = await puppeteer.launch({
            headless: 'new',
        })
    
        const url = "https://www.google.com/maps"
    
        const page = await browser.newPage()
        await page.goto(url)

        // Make new object excel
        
        const workbook = new ExcelJS.Workbook();
        await workbook.xlsx.readFile('./hotel-list/hotel1-address.xlsx');
        const worksheet = workbook.getWorksheet(1);

        const columnA = worksheet.getColumn('B').values.slice(2);

        // Wait for search selector
        await page.waitForSelector('input#searchboxinput')
        // // const input = await page.$('input#searchboxinput')
        const data = []

        let hotelCount = 0;
        // const defaultDelay = 1000;
        for(const name of columnA){
            try {
                console.log(`Processing hotel ${hotelCount + 1}: ${name}`);
                hotelCount++; 
                await page.click('input#searchboxinput', {clickCount: 3})

                await page.type('input#searchboxinput', name, {delay: 50})
                await page.keyboard.press('Enter')

                await page.waitForTimeout(8000)
                // await page.waitForNavigation({waitUntil: 'domcontentloaded'})

                // Extract data from the page
                const result = await page.evaluate((name) => {
                    const phoneElement = document.querySelector('div:nth-child(5) > button > div > div.rogA2c > div.Io6YTe.fontBodyMedium.kR99db');
                    const websiteElement = document.querySelector('a.CsEnBe');
                    
                    const phone = phoneElement ? phoneElement.textContent?.trim() : '';
                    const website = websiteElement ? websiteElement.href : '';
                    return { name, phone, website };
                }, name);

                data.push(result)
                console.log(result)
                // console.log(result)
            } catch (error) {
                console.log(error)
            }
        }   
        const outputDirectory = "./"; // Relative path to the 'coffe' directory
        const currentModuleURL = import.meta.url;
        const currentModulePath = new URL(currentModuleURL).pathname;
        const outputFilePath = path.join(path.dirname(currentModulePath), outputDirectory, 'maps-hotel1.xlsx');

        // Save the data for this subCategory to an Excel file
        const ws = XLSX.utils.json_to_sheet(data);
        const wb = XLSX.utils.book_new();
        XLSX.utils.book_append_sheet(wb, ws, "Sheet1");
        XLSX.writeFile(wb, outputFilePath);
    
        await browser.close()    
    } catch (error) {
        console.log(error)
    }
})()

File: map.py
import time
from selenium import webdriver
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.common.by import By
from selenium.common.exceptions import NoSuchElementException
from openpyxl import load_workbook
import os

def scrape_hotels():
    try:
        # Configure the ChromeOptions
        chrome_options = webdriver.ChromeOptions()
        chrome_options.add_argument('--headless')  # To run Chrome in headless mode

        # Start the WebDriver
        driver = webdriver.Chrome(options=chrome_options)

        url = "https://www.google.com/maps"
        driver.get(url)

        # Make new object excel
        workbook = load_workbook('./hotel-list/hotel1-address.xlsx')
        worksheet = workbook.active

        column_b_values = [cell.value for cell in worksheet['B'][1:]]

        # Wait for search selector
        search_input = driver.find_element(By.XPATH, '//input[@id="searchboxinput"]')

        data = []
        hotel_count = 0
        default_delay = 1  # seconds

        for name in column_b_values:
            try:
                print(f"Processing hotel {hotel_count + 1}: {name}")
                hotel_count += 1

                # Click on the search input and clear any existing text
                search_input.click()
                search_input.clear()

                # Type the hotel name and press Enter
                search_input.send_keys(name)
                search_input.send_keys(Keys.ENTER)

                time.sleep(default_delay * 10)  # Wait for the page to load (adjusted to 10 seconds)

                # Extract data from the page
                try:
                    phone_element = driver.find_element(By.XPATH, '//div[@class="rogA2c"]')
                    phone = phone_element.text.strip()
                except NoSuchElementException:
                    phone = ''

                try:
                    website_element = driver.find_element(By.XPATH, '//a[@class="CsEnBe"]')
                    website = website_element.get_attribute('href')
                except NoSuchElementException:
                    website = ''

                result = {'name': name, 'phone': phone, 'website': website}
                data.append(result)
                print(result)
            except Exception as error:
                print(error)

        output_directory = "./"
        output_file_path = os.path.join(output_directory, 'maps-hotel1.xlsx')

        # Save the data to an Excel file
        workbook.save(output_file_path)

        driver.quit()

    except Exception as error:
        print(error)

if __name__ == "__main__":
    scrape_hotels()
